Add kmir tool description and CI workflow

jberthold opened the pull request:
This PR adds the `kmir` tool to the doc.s and sets up a CI workflow that runs the tool on a few examples.
The examples are drawn from Challenge 11 (unchecked arithmetic operations). 

The tool is described using the text from the related issue #296 , and the example proofs come with a description that explains how they were set up.

The `kmir` tool is work in progress, we expect tool usage to change. Specifically the proof setup will be automated in a future version, enabling the tool to work directly with Rust code.

`kmir` is provided as a docker image from Dockerhub for the CI workflow. This image also includes `stable-mir-json` and was assembled for general usage by developers to try the tool in its current state.

Resolves #296

jberthold committed:
add KMIR to the docs using prepared tool template text

jberthold committed:
add kmir workflow and proof claim files + source. WIP README.mds need adjustments

jberthold committed:
remove stray quotes

jberthold committed:
New README.md for the KMIR proof top directory

jberthold committed:
Adjust README for maximum-example-proof

jberthold committed:
adjust steps in maximum proof, give an example of the configuration

jberthold committed:
use demon container in workflow

jberthold committed:
Do not set a default shell in the CI workflow

When setting `shell: bash`, the internal command used in the github job is different from when nothing is set, see https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_iddefaultsrunshell

jberthold committed:
Do not set uid, do not bind-mount (copy kmir-proofs directory)

jberthold committed:
do not write proof kcfg in CI (works around a permission problem)

jberthold committed:
Merge branch 'main' into add-kmir-tool

F-WRunTime committed:
Update CI Image w/ new environment controls 

- Focused on fixing read/write permission issues in previous image
- New image introduces new env and bash profile for additional exec calls used in CI.

F-WRunTime committed:
Adding back in proof generation on container

- Still requires some additional steps to fetch the results from the running container but this is a starting point.

F-WRunTime committed:
Revert previous change to workflow. Further investigation needed between CI environments

jberthold committed:
Merge branch 'main' into add-kmir-tool

jberthold commented:
@tautschnig Thank you very much for your detailed feedback on the text. Will try to address your points in a revised version coming soon.

> 1. Meta-note: very long lines make it quite hard to given focussed feedback. Mind breaking lines at 80, 10, or maybe 120 characters?

Very true, sorry for this. Keeping each paragraph in a single line is sometimes necessary for weak renderers but certainly not a good idea for a PR to be commented on.

In a first commit from here, I will just reformat all Markdown to 80-column paragraphs (w/o any non-formatting changes) to get this problem out of the way.

jberthold committed:
Format Markdown files to 80-column line breaks (content unchanged)

jberthold commented on .github/workflows/kmir.yml:
```diff
@@ -0,0 +1,45 @@
+          docker run --detach --rm -t \
+            --name ${{ env.container_name }} \
+            -w /workspace \
+            runtimeverificationinc/kmir:ubuntu-jammy-0.3.112_7.1.229-9e17ccd /bin/bash
```
The image is based on [the `Dockerfile.kmir` in `mir-semantics`](https://github.com/runtimeverification/mir-semantics/blob/c221c81d73a56c48692a087a2ced29917de99246/Dockerfile.kmir), and contains a K installation, the `kmir` tool (installed and on the default path of the image user)  as well as `stable-mir-json`, a custom `rustc` which extracts Stable MIR as JSON or as graphviz' `*.dot` when compiling a crate. I'll add that to the description with a link to the Dockerfile.

jberthold committed:
Edits round 1, see related issue

dkcumming committed:
Usage instructions list current build steps only

dkcumming committed:
added missing link to proofs

dkcumming committed:
removed speculation about kup integration

dkcumming committed:
Added quote of what K Framewwork is

jberthold committed:
break new text at column 80, remove trailing whitespace

jberthold committed:
Rewrite usage section, delete CI and Versioning parts, move Licensing

jberthold committed:
expand K framework explanation to explain reachability proofs

jberthold commented:
Here is a revised version of the `kmir` description, where we have hopefully addressed all your comments and questions.
* More elaboration on how K Framework works in general (relating it to `kmir`), explaining the proof by all-path reachability through symbolic execution.
* removed parts about future `kup` based installation
* mentioned and explain the docker image we currently recommend
* removed some sections from the tool application template which seem out of place in the description
*  rewrite of how to use the tool, reflecting current state

`kmir` is in beta and under active development so I am afraid the "Steps to use" section will have to be adapted as soon as future functionality becomes available. We will do this while also adapting the existing proofs to use available automation going forward.

jberthold committed:
Merge branch 'main' into add-kmir-tool

jberthold committed:
Merge branch 'main' into add-kmir-tool

jberthold committed:
add a section explaining how loops and recursion can be addressed by K

jberthold committed:
Merge branch 'main' into add-kmir-tool

tautschnig commented:
Thank you for the updated description! What is still unclear to me is soundness of verification:
1. Do you have a list of known issues/limitations? (This applies even to bounded, sequential programs.)
2. When faced with a program that involves unboundedness of some sort (an inductive data structure required as input, a loop with no known bound), what will happen? From the latest changes I understand that at least the loop aspect of this can be supported via loop invariants, but what if a) no loop invariant is provided or b) the provided invariant is not actually an inductive invariant?

Thanks!

gregorymakodzeba committed:
Update kmir.md

Added description clarifying KMIR's approach to handling undefined behavior (UB) in Rust’s MIR. This section explains KMIR’s refusal-to-execute strategy and how UB is handled during symbolic execution to ensure soundness guarantees.

tautschnig commented on kmir-proofs/maximum-example-proof/README.md:
```diff
@@ -0,0 +1,194 @@
+
+## Running the prover on the claim and viewing the proof
+
+Now that we have constructed claim, we can run use the KMIR verifier
```
```suggestion
Now that we have constructed claim, we can use the KMIR verifier
```

tautschnig commented:
Given that the proof construction is, thus far, manual: what does it take to actually prove any standard library functions? Even it it is just a function that we already have a proof for via any of the other three tools: I think we strictly need this CI workflow to actually prove some function(s) in the standard library instead of just running a couple of examples.

dkcumming committed:
Update kmir-proofs/maximum-example-proof/README.md

Co-authored-by: Michael Tautschnig <mt@debian.org>

jberthold committed:
Merge branch 'main' into add-kmir-tool

jberthold committed:
Merge remote-tracking branch 'upstream/main' into add-kmir-tool

jberthold committed:
update unchecked_arithmetic proofs (new runner, new test code, new workflow steps)

jberthold committed:
remove maximum-example-proof (outdated)

jberthold committed:
rewrite description of proofs in kmir-proofs

jberthold committed:
Adjust kmir.md description to updated software

jberthold commented:
> Given that the proof construction is, thus far, manual: what does it take to actually prove any standard library functions? Even it it is just a function that we already have a proof for via any of the other three tools: I think we strictly need this CI workflow to actually prove some function(s) in the standard library instead of just running a couple of examples.

Hi @tautschnig , we made some good progress in the past two weeks and I updated the `kmir` description and the example proofs (and description) to reflect that.

Please take a look at [the new desciption](https://github.com/model-checking/verify-rust-std/blob/d06457a4657386cadcaa7c7b719a0b7f3a4c2d17/kmir-proofs/README.md#example-proving-absense-of-undefined-behaviour-in-unchecked_-arithmetic).
With the new version, `kmir prove-rs` operates on actual Rust code - test functions that state properties as pre-conditions (conditional execution) and assertions (of result properties) in a harness around the Rust code under test (specifically calls to `stdlib` functions that we are proving).
I hope this fits your requirement of "actually prove[ing] any standard library functions".  

The new description in `kmir.md` is relatively short because some details of how to operate the proofs might still change and I would like the description to stay current. The `README.md` in `kmir-proofs` has more detail. However, if you prefer a longer description in `kmir.md`  I can also move the `kmir-proofs/README.md` into `kmir.md`.

jberthold committed:
Merge branch 'main' into add-kmir-tool

dkcumming committed:
Rename Stable MIR to Public MIR

dkcumming committed:
Updated installation and usage sections

dkcumming committed:
Added section on useful proving flags

dkcumming committed:
s/symbollic/symbolic & s/Solitidy/Solidity

dkcumming committed:
Added installation gif

dkcumming committed:
Added passing proof and show gif

dkcumming committed:
Added failing proof with view gif

dkcumming committed:
Added some examples and clarity for loops.

dkcumming committed:
Added section on KMIR proof harnesses

dkcumming committed:
Added case study for Challenge 11

dkcumming committed:
Added case study for P-Token proofs

dkcumming committed:
Added known limitations sections

dkcumming committed:
Use `ubuntu-jammy-latest` as docker release in workflow

dkcumming committed:
Added challenge 11 proofs to workflow

Removed old unchecked_arithmetic proofs

dkcumming commented:
@tautschnig I have updated the application. I was wondering if you wanted me to make a merge commit on top here, or to do a rebase?

dkcumming committed:
Updated `run-proofs.sh` to reuse `llvm-kompile`

Multiple start-symbols per filee mean time can be saved on kompilation.
Increases overall performance by approx 3x.

dkcumming committed:
Added parallel execution of 2 for run-proofs.sh

2 should be fine for a public github runner. I attempted to interleave
the proofs that are different so that the expensive `llvm-kompile` calls
will happen at different times.

dkcumming committed:
Merge remote-tracking branch 'upstream/main' into add-kmir-tool

dkcumming commented:
Hey @tautschnig, it would be good to get some feedback on this. CI is passing.

dkcumming committed:
Added negative tests to demonstrate KMIR errors on UB path

dkcumming committed:
Adding proofs negative harnesses (forgotton last commit)

dkcumming committed:
Merge branch 'main' into add-kmir-tool

dkcumming commented:
@feliperodri hope it is okay to ping unannounced, I wonder if you might be a better candidate to get feedback from

feliperodri commented:
@dkcumming this PR is marked as draft, that's why I didn't review it before. Is it ready for a review? I saw in the GitHub issue associated with this application that the tool application already received the go ahead. Turn this into non-draft if it's ready for a full review.

feliperodri commented:
@dkcumming also resolve any comment that you already addressed. This will facilitate the review and I won't need to double check work that has been already done.

dkcumming commented:
@feliperodri okay sorry I think I got the process misunderstood I will do those things and am in the process of correcting the other PRs you reviewed on Sunday. I intend to correct these things earnestly over the next couple of days. Thank you!

dkcumming commented on doc/src/tools/kmir.md:
```diff
@@ -0,0 +1,90 @@
+**KMIR**
+
+## Description
+[KMIR](https://github.com/runtimeverification/mir-semantics) is a formal verification tool for Rust that defines the operational semantics of Rust’s Middle Intermediate Representation (MIR) in K (through Stable MIR). By leveraging the [K framework](https://kframework.org/), KMIR provides a parser, interpreter, and symbolic execution engine for MIR programs. This tool enables direct execution of c ...
```
1. All lines split at 80 chars in [5746c4d](https://github.com/runtimeverification/verify-rust-std/commit/5746c4db8a6705311fdd3fa4a025c28fbb18fa01) and [9feeb14](https://github.com/runtimeverification/verify-rust-std/commit/9feeb14d694286f8669567ccbf1368137ebb7420);
2. Removed speculation about `kup` in [08de801](https://github.com/runtimeverification/verify-rust-std/commit/08de8017ebf7efe4e4aad8c35a0dddd37161fba0) and added installation instructions that use `kup` in [e4377e7](https://github.com/runtimeverification/verify-rust-std/commit/e4377e7a8e4647e8ad8df5205c9a5f3c8d5d0a6e);
3. Further explanation of proofs with K Framework and examples in [e62f890](https://github.com/model-checking/verify-rust-std/pull/310/commits/e62f890dffc58728312d05c0b6165e193011e46a), [7b857e9](https://github.com/model-checking/verify-rust-std/pull/310/changes/7b857e9a98843b4e1206ef1c6161c7fdbac07245), [b7d54d9](https://github.com/model-checking/verify-rust-std/pull/310/changes/b7d54d9b72c67b7e1c8c5fbc7711710a51dd064b), [39a45e2](https://github.com/model-checking/verify-rust-std/pull/310/changes/39a45e2bc9d97af997681a27eea3859b3ebaa9ab), [8632482](https://github.com/model-checking/verify-rust-std/pull/310/changes/86324826ffe16ba0f321b1bf60b4225f1bcbdb57).

dkcumming commented on doc/src/tools/kmir.md:
```diff
@@ -0,0 +1,90 @@
+![kmir_env_diagram_march_2025](https://github.com/user-attachments/assets/bf426c8d-f241-4ad6-8cb2-86ca06d8d15b)
+
+
+Particular to this challenge, KMIR verifies program correctness using the correct-by-construction symbolic execution engine and verifier derived from the K encoding of the Stable MIR semantics. The K semantics framework is based on reachability logic, which is a theory describing transition systems in [matching logic](http://www.matching-logic.org/). Transition rules of the semantics are rewritin ...
```
- Points 1. and 2. are addressed in [d68fec5](https://github.com/model-checking/verify-rust-std/pull/310/commits/d68fec5ddcb869622fd50585958ed9548477220a);
- Point 3. is restated in [7b857e9](https://github.com/model-checking/verify-rust-std/pull/310/changes/7b857e9a98843b4e1206ef1c6161c7fdbac07245);
- Point 4. addressed in [e62f890](https://github.com/model-checking/verify-rust-std/pull/310/changes/e62f890dffc58728312d05c0b6165e193011e46a).

dkcumming commented on doc/src/tools/kmir.md:
```diff
@@ -0,0 +1,90 @@
+## Tool Information
+
+* [x] Does the tool perform Rust verification? 
+  *Yes – It performs verification at the MIR level, which is a critical intermediate representation of Rust programs.*
```
Addressed in [3acce27](https://github.com/model-checking/verify-rust-std/pull/310/changes/3acce27d48830c09d9859076e1e6a20a542df669)

dkcumming commented on doc/src/tools/kmir.md:
```diff
@@ -0,0 +1,90 @@
+* [x] Does the tool deal with *unsafe* Rust code? 
+  *Yes – By operating on MIR, KMIR can analyze both safe and unsafe Rust code.*
+* [x] Does the tool run independently in CI? 
+  *Yes – KMIR can be integrated into CI workflows via our package manager and Nix-based build system.*
```
Addressed in [3acce27](https://github.com/model-checking/verify-rust-std/pull/310/changes/3acce27d48830c09d9859076e1e6a20a542df669) and [e4377e7](https://github.com/model-checking/verify-rust-std/pull/310/changes/e4377e7a8e4647e8ad8df5205c9a5f3c8d5d0a6e)

dkcumming commented on doc/src/tools/kmir.md:
```diff
@@ -0,0 +1,90 @@
+## Comparison to Other Approved Tools
+The other tools approved at the time of writing are Kani, Verifast, and Goto-transcoder (ESBMC).
+
+- **Verification Backend:** KMIR primarily differs from all of these tools by utilizing a unique verification backend through the K framework and reachability logic (as explained in the description above). KMIR has little dependence on an SAT solver or SMT solver. Kani's CBMC backend and Goto-transcode (ESBMC) encode the verification problem into an SAT / SMT verification condition to be discharg ...
```
Addressed in [3acce27](https://github.com/model-checking/verify-rust-std/pull/310/changes/3acce27d48830c09d9859076e1e6a20a542df669)

dkcumming commented on doc/src/tools/kmir.md:
```diff
@@ -0,0 +1,90 @@
+* [x] Does the tool run independently in CI? 
+  *Yes – KMIR can be integrated into CI workflows via our package manager and Nix-based build system.*
+* [x] Is the tool open source?
+  *Yes – KMIR is open source and available on GitHub.*
```
Addressed in [3acce27](https://github.com/model-checking/verify-rust-std/pull/310/changes/3acce27d48830c09d9859076e1e6a20a542df669)

dkcumming commented on doc/src/tools/kmir.md:
```diff
@@ -0,0 +1,90 @@
+* [x] Is the tool open source?
+  *Yes – KMIR is open source and available on GitHub.*
+* [x] Is the tool under development? 
+  *Yes – KMIR is actively under development, with ongoing improvements to MIR syntax coverage and verification capabilities.*
```
Addressed in [c2953b4](https://github.com/model-checking/verify-rust-std/pull/310/commits/c2953b4cd94562a9473ff9fcc93fd98c0e30bc8a)

dkcumming commented:
@feliperodri I think I need write access to model-checking/verify-rust-std in order to resolve comments and mark this as ready for review. I added links on all comments to the relevant commits, and if I could resolve them I would. What should be do, is there some step I am missing to have that ability or is someone with write access able to do that for me? The original PR was opened by my colleague who is no longer working at RV, maybe that is related.

feliperodri reviewed (approved):
Reviewed the full PR (doc page, workflow, and the Challenge 11 harnesses). This is a clean, additive tool onboarding — no changes to `library/` and nothing that touches Kani or any existing tool/CI — and both KMIR jobs are green on the current head, so the Docker-based workflow is demonstrably functional. Approving.

The positive harnesses correctly guard each `unchecked_*` op behind the matching `checked_*` precondition, and the negative harnesses (with `run-proofs.sh --negative` inverting the exit code) are the right shape for asserting UB is detected. LGTM.

A few non-blocking follow-ups:

1. **Pin the Docker image tag.** `kmir.yml` uses `runtimeverificationinc/kmir:ubuntu-jammy-latest`, but the docs recommend pinning to a specific version (e.g. `ubuntu-jammy-0.4.206`). An unpinned `latest` can silently change and break CI later without a corresponding commit — pinning it keeps the workflow reproducible.

2. **Tighten the `push:` trigger.** The `push` trigger has no branch filter and includes `library/**` in its paths, so KMIR CI fires on any push touching std (which the KMIR proofs never read) and duplicates the `pull_request` run. Consider dropping `library/**` from the push paths and/or restricting to `main`.

3. **Doc polish.** A few phrasings read a bit stronger than the current tool maturity ("provably complete", "sound assurances"), and there are minor trailing-whitespace / trailing-blank-line nits (e.g. the extra blank lines appended to `doc/src/tools.md`). Worth a light editorial pass together with the wording comments already raised in-thread, but not a merge blocker.

feliperodri commented:
> @feliperodri I think I need write access to model-checking/verify-rust-std in order to resolve comments and mark this as ready for review. I added links on all comments to the relevant commits, and if I could resolve them I would. What should be do, is there some step I am missing to have that ability or is someone with write access able to do that for me? The original PR was opened by my colleague who is no longer working at RV, maybe that is related.

I think the problem is because your colleague opened the PR (sorry for taking so long to review it). It looks good to me and I'll get another committee member to give another review and we should be good to go. You can address all remaining comments/suggestions on follow up PRs and you can track these improvements creating GitHub issues. Thanks for this contribution!

dkcumming commented:
@feliperodri I am happy to address those comments immediately. But if that would interfere with the second review I can hold off. Let me know what you prefer.

feliperodri commented:
@dkcumming that would not interfere with the second review. You can go ahead and implement the comments.

dkcumming committed:
Fixed version of KMIR on CI and updated docker instructions in README

dkcumming committed:
Tightened push triggers

dkcumming committed:
Editorial clean up and weakening of statements

dkcumming committed:
Updated and edited stale kmir-proofs/README.md

dkcumming committed:
Updated Cargo.lock

dkcumming commented:
@feliperodri I implemented the changes you requested! Thank you for your review.

feliperodri commented:
> @feliperodri I implemented the changes you requested! Thank you for your review.

Thank you, @dkcumming! @HuStmpHrrr this is ready for a final review.

HuStmpHrrr commented on kmir-proofs/0011-floats-ints/carrying_mul.rs:
```diff
@@ -0,0 +1,36 @@
+#![feature(bigint_helper_methods)]
+
+fn main() {
+    carrying_mul_u8(0, 0, 0);
```
it seems all inputs are zero. does it only check zero inputs, or does it check all inputs?

dkcumming replied:
@HuStmpHrrr these calls are added to `main` so that the compilation will not optimise out the functions; the proofs on CI never run `main`. [Usage (Verification)](https://github.com/model-checking/verify-rust-std/pull/310/changes#diff-54d45643eacaaab13b243cb705a0fab94c28783fa3b8188e41e7f0066797d0d2R316-R318) of `doc/src/tools/kmir.md` explains that if the start symbol is provided explicitly it is the entrypoint:

>Where `<FILE>` is the Rust source file to verify, `<DIR>` is the directory
where proof artifacts are stored, and `<SYMBOL>` is the entry function to
verify (defaults to `main` if omitted).

For `carrying_mul.rs` those symbols are passed in [run-proofs.sh:38](https://github.com/model-checking/verify-rust-std/pull/310/changes#diff-3c6084423b494900a696b4f4a80154c38fa61d65ab19b30e6af3f6c1e6062539R38)

> "kmir prove carrying_mul.rs  --terminate-on-thunk --proof-dir $PROOF_DIR --start-symbol carrying_mul_u8,carrying_mul_u16,carrying_mul_u32,carrying_mul_u64" \

Do those sections help enough to understand, or should I specifically mention the inclusion of the functions in `main` to stop them being optimised out?

HuStmpHrrr replied:
I see now I understand. I think some clarity is good to have. otherwise, I don't have more feedbacks to provide.
